### PR TITLE
Add support for cmake build system

### PR DESCRIPTION
**Pull request by DusanJovic-NOAA**

### CODE REVIEW

File: modulefiles/gaea.intel/fv3.intel-16.0.3.210 [hunk 1]
```diff
@@ -41,3 +41,14 @@ module load esmf/8.0.0_bs50
 
 # Needed at runtime:
 module load alps
+
```
Comment (climbfuji) [inline modulefiles/gaea.intel/fv3.intel-16.0.3.210:44]: Shouldn't this block be the same as in modulefiles/gaea.intel/fv3?

File: modulefiles/gaea.intel/fv3 [hunk 2]
```diff
@@ -39,3 +39,12 @@ module load esmf/8.0.0
...
 module load alps
+
+##
+## load cmake
+##
+setenv CMAKE_C_COMPILER cc
+setenv CMAKE_CXX_COMPILER CC
+setenv CMAKE_Fortran_COMPILER ftn
+setenv CMAKE_Platform gaea
+setenv NETCDF ${NETCDF_DIR}
```
Comment (climbfuji) [inline modulefiles/gaea.intel/fv3:50]: Do we need a line `setenv CMAKE_Platform gaea.intel` here?

**Review by climbfuji — COMMENTED** [review]
This is only a partial review, I am still working on it / testing. But I wanted to let you know where I am. Besides the two comments below, I am getting compilation errors for the IPD cmake version on macOS+GNU with the following flags:
```
./compile_cmake.sh $PWD/.. macosx.gnu 'OPENMP=N 32BIT=Y' '32bit' 2>&1 | tee compile_ipd_new.log
```
The errors are of this type:
```
/Users/heinzeller-d/TMP_NOAA/scratch/ufs-weather-model/ufs-weather-model-cmake-rt-for-develop-branch-20191208/compile/ipd_new/FV3/gfsphysics/physics/aer_cloud.F:3173:13:

 3173 |        delw0=cubicint_ice(SW, SW0, 1d0, 0d0, 1d0)
      |             1
Error: Type mismatch in argument 'y2' at (1); passed REAL(16) to REAL(8)
```
I hope I have time to fix this tonight and propose changes. This problem doesn't occur for CCPP, the following compile commands I do get an executable (didn't run it yet; note also that the OPENMP=N is required on the particular macbook I am currently on, it runs the latest macOS Catalina and this version messed up the OpenMP libraries):
```
./compile.sh $PWD/../FV3 macosx.gnu 'OPENMP=N SION=Y 32BIT=Y' '32bit' 2>&1 | tee compile_ipd_old.log
./compile.sh $PWD/../FV3 macosx.gnu 'OPENMP=N SION=Y 32BIT=Y CCPP=Y' '32bit' 2>&1 | tee compile_ccpp_dynamic_old.log
./compile.sh $PWD/../FV3 macosx.gnu 'OPENMP=N SION=Y 32BIT=Y CCPP=Y STATIC=Y SUITES=FV3_GFS_2017_gfdlmp,FV3_GSD_v0' '32bit' 2>&1 | tee compile_ccpp_static_old.log
./compile_cmake.sh $PWD/.. macosx.gnu 'OPENMP=N 32BIT=Y CCPP=Y STATIC=Y SUITES=FV3_GFS_2017_gfdlmp,FV3_GSD_v0' '32bit' 2>&1 | tee compile_ccpp_static_new.log
```
The cmake CCPP dynamic new fails as expected (because we don't support it):
```
./compile_cmake.sh $PWD/.. macosx.gnu 'OPENMP=N 32BIT=Y CCPP=Y' '32bit' 2>&1 | tee compile_ccpp_dynamic_new.log
```
I will also try the 64bit versions tonight, as well as DEBUG and REPRO mode, and run all possible executables for at least one setup (GFDL-MP basic setup).

**Comment by climbfuji** [comment]
@DusanJovic-NOAA can you please apply this patch to the fv3atm submodule?
```
diff --git a/gfsphysics/CMakeLists.txt b/gfsphysics/CMakeLists.txt
index d2814a0..6c2a236 100644
--- a/gfsphysics/CMakeLists.txt
+++ b/gfsphysics/CMakeLists.txt
@@ -10,7 +10,7 @@ if(CMAKE_Fortran_COMPILER_ID MATCHES "Intel")
         string (REPLACE "-i4 -real-size 32" "-i4 -real-size 64 -no-prec-div -no-prec-sqrt" CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS}")
     endif()
 elseif(CMAKE_Fortran_COMPILER_ID MATCHES "GNU")
-    set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fdefault-real-8")
+    set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fdefault-real-8 -fdefault-double-8")
 endif()
 endif()
```
With this, the compile_cmake.sh works on macOS+GNU for IPD as well. Thanks!

Also: what do you think of changing
```
make -j ${MAKE_THREADS}
```
in `compile_cmake.sh` to
```
make VERBOSE=1 -j ${MAKE_THREADS}
```
automatically if `DEBUG=Y` is in the `MAKEOPT` string?

**Comment by DusanJovic-NOAA** [comment]
Added "fdefault-double-8" to gfsphysics/CMakeLists.txt in b0c4740.
We can add VERBOSE flag to make for debug builds, but I suggest we make that change in one of the future commits, simply to avoid rerunning full regression tests on all platforms.

**Comment by climbfuji** [comment]
> Added "fdefault-double-8" to gfsphysics/CMakeLists.txt in [b0c4740](https://github.com/ufs-community/ufs-weather-model/commit/b0c4740315934ab0ae57a15d360df5d6ee6be72c).
> We can add VERBOSE flag to make for debug builds, but I suggest we make that change in one of the future commits, simply to avoid rerunning full regression tests on all platforms.

Thanks for making the change. Agreed, VERBOSE is not a pressing change to make.

**Review by climbfuji — APPROVED** [review]
I tested this PR on macOS with both gcc+gfortran and clang+gfortran for a range of MAKEOPTS (32bit, 64bit) in DEBUG mode without OpenMP for IPD, CCPP dynamic, CCPP static, both old make and new cmake build. All versions I tested thus far compile and give b4b identical results between IPD and the two CCPP builds, also between gcc+gfortran and clang+gfortran. While I will keep testing with OpenMP and in REPRO and PROD mode, I think this PR is good to go.

Edit: the test I ran were the basic fv3_gfdlmp/fv3_ccpp_gfdlmp regression tests.

**Review by junwang-noaa — APPROVED** [review]

**Comment by climbfuji** [comment]
Once this is merged, I will update dtc/develop and start to prepare for the commit to develop/master for Dec 20.

I did more tests with OpenMP being turned on on macOS for gcc+gfortran, and the results look good.

**Comment by junwang-noaa** [comment]
Dom,

So you plan to combine these changes with CCPP changes for coupled system?

**Comment by climbfuji** [comment]
> Dom, So you plan to combine these changes with CCPP changes for coupled system?
> […](#)

Yes, I owe you a detailed description of what the PR contains, and I will do this as I create the PRs tomorrow and probably Thursday.

**Review by RatkoVasic-NOAA — APPROVED** [review]
Looks good to me.

**Pull request merged**